Add basic multitouch support for osu!

o-dasher opened the pull request:
https://user-images.githubusercontent.com/88356162/186967880-0413298b-fd41-48b3-afcf-113e8fff3dc8.mp4

This is a simple implementation of osu! ruleset touch support, to make this implementation i followed the specifications at #18882, this could still be improved by considering the cursor touch to be the cursor closer to the current hittable object at the playfield, but this work is probably way too big for a single pull request.

Commit:
Simple touchscreen std implementation

Commit:
Minor code quality improvement

Commit:
Adds comment explaining dragmode

Commit:
Fix code quality

Commit:
simplifies select active actions expression

Commit:
Rename wrongly named variable

Commit:
Minor code quality improvement

Commit:
Fixes o(n) operation

Commit:
Minor optimization

Commit:
Make code clearer

frenzibyte commented on osu.Game.Rulesets.Osu/OsuInputManager.cs:
```diff
@@ -43,17 +53,52 @@ protected override bool Handle(UIEvent e)
         protected override bool HandleMouseTouchStateChange(TouchStateChangeEvent e)
         {
-            if (!AllowUserCursorMovement)
+            var source = e.Touch.Source;
+
+            int touchNumber = numberedTouchSources[source];
+
+            if (touchNumber > simultaneous_touches_limit)
+                return false;
```
`TouchSource` is made in a way that allows mapping sources to integers without the need of any extra object or explicit dictionaries.
```suggestion
            var source = e.Touch.Source;
            if (source - TouchSource.Touch1 >= simultaneous_touches_limit)
                return false;
```

frenzibyte commented on osu.Game.Rulesets.Osu/OsuInputManager.cs:
```diff
@@ -43,17 +53,52 @@ protected override bool Handle(UIEvent e)
             return base.Handle(e);
         }
 
+        private OsuAction getActionForTouchSource(TouchSource source)
+        {
+            int sourceNumber = numberedTouchSources[source];
+            return sourceNumber % 2 == 0 ? OsuAction.RightButton : OsuAction.LeftButton;
+        }
```
Same as above, `numberedTouchSources` is not required.
```suggestion
        private OsuAction getActionForTouchSource(TouchSource source) => (int)source % 2 == 0 ? OsuAction.RightButton : OsuAction.LeftButton;
```

frenzibyte commented on osu.Game.Rulesets.Osu/OsuInputManager.cs:
```diff
@@ -43,17 +53,52 @@ protected override bool Handle(UIEvent e)
+
+            if (isTapTouch)
+            {
+                var action = getActionForTouchSource(source);
```
For consistency with this code, the action should be determined by the index of the touch in the active tap touches.
```suggestion
                int tapNumber = activeTapTouches.TakeWhile(t => t != source).Count();
                var action = tapNumber % 2 == 1 ? OsuAction.RightButton : OsuAction.LeftButton;

```

frenzibyte commented on osu.Game.Rulesets.Osu/OsuInputManager.cs:
```diff
@@ -43,17 +53,52 @@ protected override bool Handle(UIEvent e)
+            if (touchNumber > simultaneous_touches_limit)
+                return false;
+
+            var activeSources = CurrentState.Touch.ActiveSources;
+
+            var cursorTouch = activeSources.Any() ? activeSources.First() : source;
+
+            var activeTapTouches = AllowUserCursorMovement ? activeSources.Skip(1) : activeSources;
+            var limitedActiveTapTouches = activeTapTouches.Take(tap_touches_limit);
+
+            var newActiveTapActions = limitedActiveTapTouches.Select(getActionForTouchSource);
+            var newInactiveTapActions = PressedActions.Where(a => !newActiveTapActions.Contains(a)).ToList();
```
As mentioned above, this should be saved in private fields rather than reallocating every time a touch changes state.

frenzibyte commented on osu.Game.Rulesets.Osu/OsuInputManager.cs:
```diff
@@ -15,6 +17,9 @@ namespace osu.Game.Rulesets.Osu
 {
     public class OsuInputManager : RulesetInputManager<OsuAction>
     {
+        private const int tap_touches_limit = 2;
+        private const int simultaneous_touches_limit = tap_touches_limit + 1;
```
Please add xmldoc explaining the difference between "tap" and "simultaneous".

frenzibyte reviewed (changes requested):
Thanks for the contribution. I haven't tested this in mobile yet, but code-wise this needs a rework for the following reasons:
 - Key-binding actions should never be triggered inside the input propagation flow, let alone in the `HandleMouseTouchStateChange` path which is meant to propagate mouse input from touch, not handling touch.
 - There are multiple LINQ operations/allocations occurring every time a touch has changed state, this could get quite bad performance-wise.

Take a look at how touch support is added in osu!mania for inspiration, it's all isolated in a `Component` that handles touch events properly.

In addition, tests are immensely required to understand the implications of the logic you currently have here, as it doesn't look straightforward enough to not have one, and cleaning it up may potentially result in breaking actual behaviour (take a look at `OsuManualInputManagerTestScene` for sending input in a test scene and `CreateBeatmap` for defining a beatmap that's optimal for testing).

Commit:
Optimizing allocations

Commit:
Fixes Unnecessary allocations and linq

Commit:
Removes unused variables and fix tap logic

Commit:
Map tap actions into a dictionary for performance

Commit:
Simplify tap touch trigger

o-dasher commented:
> Thanks for the contribution. I haven't tested this in mobile yet, but code-wise this needs a rework for the following reasons:
>  - Key-binding actions should never be triggered inside the input propagation flow, let alone in the `HandleMouseTouchStateChange` path which is meant to propagate mouse input from touch, not handling touch.
>  - There are multiple LINQ operations/allocations occurring every time a touch has changed state, this could get quite bad performance-wise.
> 
> Take a look at how touch support is added in osu!mania for inspiration, it's all isolated in a `Component` that handles touch events properly.
> 
> In addition, tests are immensely required to understand the implications of the logic you currently have here, as it doesn't look straightforward enough to not have one, and cleaning it up may potentially result in breaking actual behaviour (take a look at `OsuManualInputManagerTestScene` for sending input in a test scene and `CreateBeatmap` for defining a beatmap that's optimal for testing).

I think i've fixed the performance concerns, i will take a look at the mania touch implementation later, thanks.

Commit:
Fixx code quality

Commit:
drawable based input implementation

Commit:
Remove whitespace

Commit:
Base testing

Commit:
Using screen space positioning

Commit:
Move test position into a step

Commit:
Actually  make input being passed to ManualInputManager

Commit:
Expose all touch sources with touch input handler

Commit:
Small refactoring

Commit:
Simple testing

Commit:
Small rewritting of AllowedTouchSources

Commit:
Fix minor performance concern

Commit:
Remove nullable anotation

Commit:
Small refactoring

o-dasher commented:
Reading through the failed tests, i don't think these are related to this branch changes, so maybe it would be necessary to merge master branch onto this, but overall i think everything seems fine to me, after a few refactors and writing the tests and re-reading the code again and again it's logic seems to be pretty straightforward.

smoogipoo commented:
I'm not really a fan of the cross communication between `OsuInputManager` and `OsuDrawableTouchInputHandler`. Since the latter is a child of the former, can't everything be implemented inside `OsuDrawableTouchHandler` itself?

o-dasher commented:
> I'm not really a fan of the cross communication between `OsuInputManager` and `OsuDrawableTouchInputHandler`. Since the latter is a child of the former, can't everything be implemented inside `OsuDrawableTouchHandler` itself?

I don't think this cross communication is avoidable, because we can't make a touch be inactive through another drawable, which is the reason to the `DragMode` exist (Disable cursor touch and make it drag only so you can click with other tap finger), and it would be inconsistent even if it was possible because of other input managers children.

Commit:
Simplifies drag mode logic

Commit:
Using lambda for inline booleans

Commit:
Fix missing lambda

peppy committed:
Clean up tests a bit (I don't know what the test is doing though)

peppy committed:
Rename weirdly named class

peppy committed:
Stop expoding `OsuTouchInputMapper`

peppy commented:
I've tidied things up for you, but there's some serious issues:

- I have no idea what the test scene is actually testing.
- It's definitely not testing enough. Reading your code I can see scenarios where it will completely fall over. Touches are not guaranteed to be in order, yet you're only reading the first x touches in the enum (the `AllowedTouches`). You will need to handle this dynamically, tracking the active touches and basing it on the number of touches down, not the enum.

peppy reviewed (changes requested):
Needs better explained (and more) test coverage.

Commit:
Fixes syntax warning

Commit:
Stop exposing allowedTouchSources in touch mapper

Commit:
Renames variables to be less confusing

Commit:
Adds xml docs explaining things a bit better

Commit:
Tracking active allowed touches

Commit:
refactors and  makes drag mode agnostic to indexes

Commit:
Merge remote-tracking branch 'upstream/master' into touch-osu

Commit:
Shortens usage of the key binding container

Commit:
Fixes test scene crash

Commit:
Makes input manager's touch mapper readonly

Commit:
Removes unused constant

o-dasher commented:
> - Touches are not guaranteed to be in order, yet you're only reading the first x touches in the enum (the `AllowedTouches`). You will need to handle this dynamically, tracking the active touches and basing it on the number of touches down, not the enum.
If you mean the "drag mode", now this part should be fixed.

Commit:
makes touch input mapper agnostic from enum

Commit:
Fixes mistaken logic for drag cursor

Commit:
Fixes typo.

Commit:
Tidy up logic to be more understandable

Commit:
Fixes typo on xmldocs

Commit:
Makes independent on touch ordering for tap touches

Commit:
Better (Non visual) test coverage

Commit:
Creates "glitchy" visual test

Commit:
Fix test touch timing

Commit:
Tidy up by using fixed test touch position

Commit:
Remove unused click

Commit:
Move variables to class start

Commit:
Stop exposing input mapper

Commit:
Better naming for tap only mapping

Commit:
Removes empty arguments

Commit:
Uses key counter for testing touch input

o-dasher commented:
> I've tidied things up for you, but there's some serious issues:
> 
> - I have no idea what the test scene is actually testing.
>   - There's no way to visualise what is going on.
>   - It's talking about dragging but the cursor isn't moving.
>   - It's not playing to the gameplay but using a player test.
>   - It's not testing any of the edge cases that would need testing.
> - It's definitely not testing enough. Reading your code I can see scenarios where it will completely fall over. Touches are not guaranteed to be in order, yet you're only reading the first x touches in the enum (the `AllowedTouches`). You will need to handle this dynamically, tracking the active touches and basing it on the number of touches down, not the enum.
I think i've addressed all these issues, including the test coverage, i think it is testing the breaking cases you had noticed now.

Commit:
Properly center test key counter

Commit:
Fix test comments

Commit:
Refactor tests into a TestFixture

Commit:
Fix cursor inactivation logic

Commit:
Fixed tests not accounting for key disabling

o-dasher commented:
I think i've found a unwanted behaviour this being that when you click with three fingers it disables the left button (this because the first touch mouse input is disabled so we can stream properly), so you have to click once again with the third finger for it to activate the left button.

Commit:
Cleanup comments

Commit:
Fixes unexpected behaviour for third finger input

Commit:
Fixes failing tests for the unexpected behaviour

Commit:
Avoid code repetition in tests

Commit:
Add  testing for third finger input counter

Commit:
Fixes code quality and uses extension properly

Commit:
Properly break line

Commit:
Tidy up test scene

Commit:
Avoids producing unnecessary touch state change events

Commit:
Testing for number of keys pressed

Commit:
Test covering currently pressed keys

Commit:
expect taps to be 0 when only one finger input

Commit:
Fix white space

Commit:
Simplifies if statements

o-dasher commented:
Autopilot touch support seems to be messed after these changes, i will try looking at it.

Commit:
Do not expose activetaptouches

Commit:
Fixes auto pilot touch input (although with some bugs)

Commit:
Fixes input limitation with autopilot mod

Commit:
Refactoring and document things further

o-dasher commented:
Ok now this is in a state that i am happy with.

Commit:
Tidy up tap touch actions into a single wrapper

Commit:
Moves checks into handle touch tap only mapping

o-dasher commented:
It seems that the code quality check fails aren't my fault so i will be ignoring these then.

peppy commented on osu.Game.Rulesets.Osu/UI/OsuTouchInputMapper.cs:
```diff
@@ -0,0 +1,143 @@
+    public class OsuTouchInputMapper : Drawable
+    {
+        /// <summary>
+        /// The default cursor touch it may fail <see cref="IsCursorTouch(TouchSource)"/> on cases where there aren't any cursor touches, such as when autopilot is on.
```
I have no idea what this is trying to say. It feels like three sentences mashed into one, all touching on different points.

o-dasher replied:
I will try to document it better when i get home, sorry.

o-dasher replied:
I believe i explained better what the `DEFAULT_CURSOR_TOUCH` means on my recent commits.

peppy commented on osu.Game.Rulesets.Osu/UI/OsuTouchInputMapper.cs:
```diff
@@ -0,0 +1,143 @@
+        /// <summary>
+        /// Tracks all active tap <see cref="TouchSource"/>s that can be mapped into a given <see cref="OsuAction"/> through the <see cref="tapTouchActions"/>.
+        /// </summary>
+        private readonly HashSet<TouchSource> activeTapTouches = new HashSet<TouchSource>();
```
Why is this required when `tapTouchActions` already exists? Seems like it's storing the same thing.

o-dasher replied:
It is required because these are the tap touch sources that were handled by the `OnTouchDown` so they can be removed and their assigned action be released later.
i though on just making a count of sctive tap touches and incrementing when a touch happens and getting the current count `OnTouchUp` and releasing the touch for the index of said count, but since tap ordering isn't guaranteed this does not work, so i believe this is a must.

o-dasher replied:
Also these active tap touches track how many tap touches are currently active so we don't accept tap input for too many fingers.

peppy commented:
I was going to review this but can't bring myself to get through it. The logic is too confusing for my brain, sorry.

Try and refactor to remove `HandleTouchTapOnlyMapping` because this crumbles everything else in the PR. It's not feasible for this to be understandable by the average developer.

o-dasher commented:
> I was going to review this but can't bring myself to get through it. The logic is too confusing for my brain, sorry.
> 
> Try and refactor to remove `HandleTouchTapOnlyMapping` because this crumbles everything else in the PR. It's not feasible for this to be understandable by the average developer.

I don't think that is possible to remove this, but i agree that it is confusing indeed, why this exists is too make the action of the first finger be disabled so the tap fingers fully take place of all the tapping done.
maybe comments explaining what everything on that method is doing should be helpful?

Commit:
Better documenting everything

Commit:
Test cover all touch sources

Commit:
Adds ouse disabled tap only mapping early-return

Commit:
Removes unnecessary overegineering

Commit:
Improves handle touch tap only mapping doc

Commit:
Rename "HandleTouchTapOnlyMapping" to be more clear

Commit:
Adds test coverage for disabled cursor movement

Commit:
Rename checkBlockCursorActions to checkBlockCursorAction

Commit:
Adds commentary explaining tap touch releasing

o-dasher commented:
> I was going to review this but can't bring myself to get through it. The logic is too confusing for my brain, sorry.
> 
> Try and refactor to remove `HandleTouchTapOnlyMapping` because this crumbles everything else in the PR. It's not feasible for this to be understandable by the average developer.

I renamed it to `BlockTouchCursorAction` which for me seems to be more readable and made comments explaining on what it is actually doing.

Commit:
Fix unused method

o-dasher committed:
Simplify comments and better naming

o-dasher committed:
Early return when cursor touch

o-dasher committed:
Fix xml doc

o-dasher committed:
Fix wronggly named assert test cases

o-dasher commented:
Hello, since releasing to ios was added to the roadmap for this month, it would be a nice addition to have atleast this base touchscreen support.
I am just asking for a review, because in a few months i may barely have any spare time to work on this anymore.
I will likely be changing to a full time school, but i am glad to help while i can with ease.

peppy commented:
I have hope of either merging this or some alternative to this. It would be great if you could fix the conflicts and do a self-check of the code in preparation for that.

o-dasher committed:
Remove merge conflicts.

o-dasher committed:
Spread out cursor touch input blocking process

o-dasher committed:
Refactors to new osu! version completely.

o-dasher committed:
Actually removes the conflict.

o-dasher committed:
Fixes failing test. Must accept the 3 touches properly.

o-dasher committed:
Adds support for autopilot

o-dasher committed:
Fixes unintended stream mode behaviour.

o-dasher committed:
Refactors touch input tests and add stream mode test

o-dasher committed:
Simplify handling on input manager

peppy committed:
Remove pointless bindable flow and properties

peppy committed:
Reorganise class and privatise `IsStreamMode` setter

peppy commented:
I started to fix this up but lost hope halfway through. I'll open a PR with my own version of this as I think it's the path of least resistance.

o-dasher commented:
Understandable, i am sorry for the commits spam, is just that i had tried many ways to solve multiple problems when trying to get the implementation right.

peppy commented:
The commit count is the least of the issue for me. The larger part is that the structure is not contained well enough. I'd hope to see very few changes to the `OsuInputManager` class, with all the logic being very localised in the touch handler instead.

o-dasher commented:
Would you be fine if i open another pull request and try to do this instead. Or would you prefer to do it yourself?

peppy commented:
I won't stop you, and while I do have relative high priority on this I do have a few higher prio things to visit first. You're more than welcome to continue work and see what you come up with. At very least I'll take inspiration from it (and potentially reuse some tests depending on their quality).

peppy commented:
@o-dasher I'll likely revisit this this week. Have you made any progress?

o-dasher commented:
> @o-dasher I'll likely revisit this this week. Have you made any progress?

I was going to get back to work on this last week although my keyboard broke, so i can't do nothing for now.

peppy commented:
I've tested this PR as is on a device and it behaves completely wrong, so I'll likely be looking at things from scratch. Here's what I was seeing in some brief testing:

When attempting to use one finger to track and two more to tap:

https://user-images.githubusercontent.com/191335/212264004-68b5c15d-fc21-4e40-9eb8-36e24c475a3a.MP4


- Works close to expectations, but buttons can get stuck in an incorrect state at times, for instance B1 generally stays lit at all times even though I'm holding nothing.

o-dasher commented:
Oh, ok then. Starting off from scratch may be better.
These errors where likely due to lack of test cases, i wasn't able to compile to android to see how stuff behaves on practice so this may had lead me to that.
I am sorry for that

peppy commented:
Fair enough. I've been able to reuse some of your code so not all is lost (at very least I used i the tests as a starting point for a new attempt).
